pkg/internal/authorizer/authorizer.go

```diff
@@ -0,0 +1,172 @@
+	}
+	return a.Client.Get(ctx, key, obj)
+}
+
+func (a AuthorizationClient) List(ctx context.Context, list runtime.Object, opts ...client.ListOption) error {
+	var authOpt ListAuthorization
+	for _, opt := range opts {
+		if namespace, isInNamespace := opt.(client.InNamespace); isInNamespace {
+			authOpt.Namespace = string(namespace)
+		}
+	}
+	decision, err := a.authorize(ctx, list, authOpt)
+	if err != nil {
+		return fmt.Errorf("authorizing request: %w", err)
+	}
+	if decision != DecisionAllowed {
+		return fmt.Errorf("not allowed: %w", err)
+	}
+	return a.Client.List(ctx, list, opts...)
+}
+
+type AuthorizationDecision string
+
+const (
+	DecisionAllowed   AuthorizationDecision = "allowed"
+	DecisionDenied    AuthorizationDecision = "denied"
+	DecisionNoOpinion AuthorizationDecision = "no opinion"
+)
+
+func (a AuthorizationClient) authorize(
+	ctx context.Context,
+	objType runtime.Object,
+	option AuthorizationOption,
+) (AuthorizationDecision, error) {
+	userInfo, err := a.getUserInfo(ctx)
+	if err != nil {
+		return DecisionDenied, fmt.Errorf("cannot get the user infomation from the request: %w", err)
+	}
+
+	gvk, err := apiutil.GVKForObject(objType, a.Scheme)
+	if err != nil {
+		return DecisionDenied, fmt.Errorf("cannot get GVK for %T: %w", objType, err)
+	}
+	var resource string
+	if meta.IsListType(objType) {
+		resource = flect.Pluralize(strings.ToLower(strings.TrimSuffix(gvk.Kind, "List")))
+
+	} else {
+		resource = flect.Pluralize(strings.ToLower(gvk.Kind))
+	}
+
+	review := &authv1.SubjectAccessReview{
+		Spec: authv1.SubjectAccessReviewSpec{
+			ResourceAttributes: &authv1.ResourceAttributes{
+				Group:    gvk.Group,
+				Version:  gvk.Version,
+				Resource: resource,
+			},
+			User: userInfo.Name,
+		},
+	}
+	option.apply(review)
+	if err := a.Create(ctx, review); err != nil {
+		return DecisionDenied, fmt.Errorf("creating SubjectAccessReview: %s", err)
+	}
+	switch {
+	case review.Status.Denied:
+		return DecisionDenied, nil
+	case review.Status.Allowed:
+		return DecisionAllowed, nil
+	default:
+		return DecisionNoOpinion, nil
+	}
+}
+
+type RequestOperation string
+
+const (
+	RequestGet  RequestOperation = "get"
+	RequestList RequestOperation = "list"
+)
+
+type AuthorizationOption interface {
+	apply(review *authv1.SubjectAccessReview)
+}
+
+type GetAuthorization struct {
+	Namespace string
+	Name      string
+}
+
+func (a GetAuthorization) apply(review *authv1.SubjectAccessReview) {
+	review.Spec.ResourceAttributes.Verb = string(RequestGet)
+	review.Spec.ResourceAttributes.Namespace = a.Namespace
+	review.Spec.ResourceAttributes.Name = a.Name
+}
+
+type ListAuthorization struct {
+	Namespace string
+}
+
+func (a ListAuthorization) apply(review *authv1.SubjectAccessReview) {
+	review.Spec.ResourceAttributes.Verb = string(RequestList)
+	review.Spec.ResourceAttributes.Namespace = a.Namespace
+}
+
+func (a AuthorizationClient) getUserInfo(ctx context.Context) (UserInfo, error) {
```

Comment:
Please don't mix member function ordering for structs. Keep it sorted.

Reply:
This function is removed because we have OIDC support